SPEC.md

```diff
@@ -0,0 +1,258 @@
+
+```js
+type UnsubscribeHandler = () => void
+```
+
+`FetchResult` is passed to the `next` callback of `Subscriber`
+
+```js
+FetchResult{
+  data?: object,
+  errors?: object[],
+  extensions?: any,
+  context?:object,
+}
+```
+
+## More examples
+
+Here is a list of planned fetchers with selected diagrams:
+
+Base:
+
+* Websocket: subscriptions support
+* Batching-\*: support for batching operations with a transport
+
+<p align="center">
+  <br>
+  <img src="images/batch-fetcher.png" alt="Batch Fetcher"/>
+</p>
+
+Intermediate:
+
+* Polling: repeats requests on a specified interval
+
+<p align="center">
+  <br>
+  <img src="images/polling-fetcher.png" alt="Polling Fetcher"/>
+</p>
+
+* Caching: returns data if result in the cache and stores data in cache on response
+* Compose: combines a list of fetchers (would require additional semantics around constructor argument order)
+* Dedup: saves query signatures that are currently on the wire and returns the result for all of those queries
+* Retry: error callback causes the request to retry
+* Mock: returns fake data for all or part of a request ← implement with BaseFetcher (eg. executes `graphql`)
+
+Fork:
+
+* Split: split operations between fetchers depending on a function passed in
+* Hybrid: uses split-fetcher to fill query and mutations with http and subscriptions with websockets
+
+<p align="center">
+  <br>
+  <img src="images/hybrid-fetcher.png" alt="Hybrid Fetcher"/>
+</p>
+
+Adapter (not a Fetcher):
+
+* Promise Wrapper
+
+<p align="center">
+  <br>
+  <img src="images/fetcher-as-promise.png" alt="Fetcher Promise Wrapper"/>
+</p>
+
+* Backwards-compatablility Wrapper: exposes `query`, `mutate`, and `subscribe`
+
+## Open questions, discussion
+
+Throughout the design process, a couple questions have surfaced that may prompt conversation. If you want to see something here, please open a PR with your proposed change or an Issue for discussion.
+
+### How should we pass context to `next` callback along with the query data?
+
+Should we pass the context from the response (for example relevant HTTP headers or status code) to the `next` callback of the observable like this:
+
+```js
+next({
+  data,
+  errors,
+  extensions,
+  context
+})
+```
+
+or like this:
+
+```js
+next({
+  result: {
+    data,
+    errors,
+    extensions
+  },
+  context
+})
+```
+
+### Should we include a status that could contain user defined data and an enum for state?
+
+```js
+enum State{
+  cold
+  started
+  stopped
+  errored
+  completed
+}
+```
+
+If `status` exists and an `Observable` has terminated, should `subscribe` throw an error or call complete immediately, avoiding memory leaks.
+
+### Does the subscribe method need to be overloaded?
+
+Currently the `subscribe` method has two different signatures, one that takes three functions `next`, `error`, and `complete` and antoher that takes a `Subscriber`.
+Positional arguments are more prone to errors than passing object containing arguments.
```

Comment:
I'm not sure I agree with this. I prefer positional arguments when the needs of the arguments are hierarchical. That being said, the spec calls for using an object so I'd go with that and not the overload https://github.com/tc39/proposal-observable#api